SI-5318 Make implicit divergence checking PolyType aware.

retronym opened the pull request:
Replaces the two active subclasses of `SymCollector`
with direct use of `Type#foreach` and `Type#typeSymbol`.

Resubmission of #589.

retronym committed:
SI-5318 Make implicit divergence checking PolyType aware.

Replaces the two active subclasses of `SymCollector`
with direct use of `Type#foreach` and `Type#typeSymbol`.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/167/

retronym commented:
review by, you guessed it, @adriaanm

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/167/

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -1401,31 +1388,21 @@ trait Infer {
+          && sym.owner.isTerm
+          && !sym.info.bounds.exists(_.typeParams.nonEmpty)
+        )
+      (for (t <- tp if includeCondition(t.typeSymbol)) yield t.typeSymbol).distinct
```
`freeTypeParamsOfTerms` will now also discover symbols behind `PolyType`-s. Is this change desired? The test case for SI-4070, mentioned above, still passes.

adriaanm replied:
yes, that sounds suspicious, since the symbols bound by a polytype are not free (by definition :))

retronym commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -346,7 +346,12 @@ trait Implicits {
       }
       def stripped(tp: Type): Type = {
-        deriveTypeWithWildcards(freeTypeParametersNoSkolems.collect(tp))(tp)
+        def isFreeTypeParamNoSkolem(sym: Symbol) = sym.isTypeParameter && sym.owner.isTerm
```
What is the purpose of `sym.owner.isTerm` here? Is it essential, or has it simply survived a series of copy-pastes since starting out in `freeTypeParamsOfTerms`?

adriaanm replied:
not sure, but it only considers method type parameters this way
it seems class type params should be stripped as well, indeed
does the spec specify?

retronym committed:
SI-5318 Avoid implicitly normalizing types in `freeTypeParamsOfTerms`.

Wildcard free class type parameters, not just method type parameters, when stripping
the core type of candidate implicits. The spec doesn't make any such distinction; and the
enclosed test, t5318c, crashes without this change.

retronym commented:
Succeeded by #625.
